add app txn visitors and struct

Blajda opened the pull request:
Implements visitors and structures required for reading application transactions that exist on a snapshot.

A similar pattern to the `LogReplayScanner` was implemented where users can explicitly scan metadata for the latest transaction for all application or obtain the latest transaction for particular application.

I decided against adding the `TransactionMap` to `Snapshot` since snapshot only require protocol + metadata to be built and obtaining the latest transaction may require scanning past that point. Also the overall design of `Snapshot` seems to be immutable and it would not make much sense to have methods that mutate it.

nicklan commented on kernel/src/transaction.rs:
```diff
@@ -0,0 +1,142 @@
+pub type TransactionMap = HashMap<String, Transaction>;
+pub struct TransactionScanner<'a> {
+    snapshot: Arc<Snapshot>,
+    engine: &'a dyn EngineInterface,
```
in general we prefer to not store the engine interface in the struct and to just pass it in where needed. this simplifies the ffi as well as making it so you don't need a lifetime on this struct.

nicklan commented on kernel/src/transaction.rs:
```diff
@@ -0,0 +1,142 @@
+        &self,
+        application_id: Option<&str>,
+    ) -> DeltaResult<TransactionMap> {
+        // Scan the entire log for all application ids but terminate early if a specific application id is provided
```
nit: this seems like a comment that belongs on the method rather than here

nicklan commented on kernel/src/transaction.rs:
```diff
@@ -0,0 +1,142 @@
+        TransactionScanner { snapshot, engine }
+    }
+
+    fn _application_transactions(
```
nit: we don't currently have a convention of naming private stuff with an `_`, so I'd just call this `application_transactions` for now

Blajda replied:
`application_transactions` already exists for this struct so I just renamed it to `scan_application_transactions`. 
When it comes to the current implementation I wanted to support obtaining the latest transaction of an application and all latest transactions. 

During implementation I had two designs in mind, either build different visitors for both use cases or just one visitor. Obtaining the latest transaction for an app id could be improved since a hash map does not need to be maintained in that instance but it would likely result in very small gains in speed / space.

nicklan reviewed (changes requested):
nice, this looks great! just a couple of small comments so far.

scovich reviewed (commented):
Thanks for taking a stab at this missing functionality!
The code generally looks good, but I think we can simplify a fair bit?

Blajda commented on kernel/src/transaction.rs:
```diff
@@ -0,0 +1,150 @@
+
+                if !self.transactions.contains_key(&txn.app_id) {
+                    self.transactions.insert(txn.app_id.clone(), txn);
+                }
```
Thanks this is succinct compare to before.

Blajda commented:
> Thanks for taking a stab at this missing functionality! The code generally looks good, but I think we can simplify a fair bit?

Thanks so much. Given your suggestions I removed the irrelevant bits and simplified the logic. From an overall unit testing perspective it would be nice if I could assert which files were read so an assert for the minimum amount of read can be performed.

scovich commented on kernel/src/actions/visitors.rs:
```diff
@@ -268,6 +270,58 @@ impl DataVisitor for RemoveVisitor {
+pub type TransactionMap = HashMap<String, Transaction>;
+
+#[derive(Default, Debug)]
+pub(crate) struct TransactionVisitor {
```
Can we add a doc comment? In particular, explaining how the visitor handles the app id filter?

scovich commented on kernel/src/actions/visitors.rs:
```diff
@@ -277,9 +331,11 @@ mod tests {
-        client::arrow_data::ArrowEngineData,
-        client::sync::{json::SyncJsonHandler, SyncEngineInterface},
+        actions::{get_log_schema, ADD_NAME, TRANSACTION_NAME},
+        client::{
+            arrow_data::ArrowEngineData,
+            sync::{json::SyncJsonHandler, SyncEngineInterface},
+        },
```
tiny nit: the deep nesting of imports can make it both bigger and harder to read:
```suggestion
        client::arrow_data::ArrowEngineData,
        client::sync::{json::SyncJsonHandler, SyncEngineInterface},
```
(a bit counter-intuitive)

scovich commented on kernel/src/transaction.rs:
```diff
@@ -0,0 +1,124 @@
+}
+
+impl TransactionScanner {
+    pub fn new(snapshot: Arc<Snapshot>) -> Self {
+        TransactionScanner { snapshot }
+    }
```
aside: @nicklan If I counted correctly, this brings us to six structs whose `new` is just a thin wrapper for convenience. Do you think it's worth bringing in the [derive-new](https://crates.io/crates/derive-new) crate? Tho this [reddit thread](https://www.reddit.com/r/rust/comments/18eln31/why_does_the_core_not_provide_a_new_derive/) has some arguments for why trivial new might be an antipattern?

nicklan replied:
Mmm, it's an interesting point. I think the only reason I'd want to keep it this way is that we could easily change things in the new method and/or add new fields with default values (which we could do using `derive-new`) and not require upstream changes.

I created #181 to remind us to think about this

scovich commented on kernel/src/transaction.rs:
```diff
@@ -0,0 +1,124 @@
+        for maybe_data in iter {
+            let (txns, _) = maybe_data?;
+            txns.extract(schema.clone(), &mut visitor)?;
+            if application_id.is_some() && visitor.transactions.len() > 0 {
```
Clippy error to fix (sorry, I didn't notice before that somebody had to approve pre-merge testing): 
```
error: length comparison to zero
  --> kernel/src/transaction.rs:36:44
   |
36 |             if application_id.is_some() && visitor.transactions.len() > 0 {
   |                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: using `!is_empty` is clearer and more explicit: `!visitor.transactions.is_empty()`
```

scovich reviewed (approved):
Very nice! 

Since it anyway needs to rebase and fix clippy errors, suggest also making a pass to add more doc and code comments so people don't have to look at this PR to understand some of the nuances that were discussed.

Blajda commented on kernel/src/transaction.rs:
```diff
@@ -26,20 +24,17 @@ impl TransactionScanner {
-        let mut visitor = TransactionVisitor {
-            transactions: HashMap::new(),
-        };
+        let mut visitor = TransactionVisitor::new(application_id.map(|s| s.to_owned()));
```
I haven't encountered any standards / conventions for this. I think it depends on the context mostly, in this case I want to convert a reference to an owned item and that aligns best with the 'why'

scovich commented on kernel/src/transaction.rs:
```diff
@@ -0,0 +1,127 @@
+                .application_transactions(&engine_interface)
+                .unwrap(),
+            txn_scan
+                .application_transaction(&engine_interface, &app_id)
```
another [clippy failure](https://github.com/delta-incubator/delta-kernel-rs/actions/runs/8862391787/job/24335971523?pr=170):
```
error: this expression creates a reference which is immediately dereferenced by the compiler
  --> kernel/src/transaction.rs:88:61
   |
88 |                 .application_transaction(&engine_interface, &app_id)
   |                                                             ^^^^^^^ help: change this to: `app_id`
```
(probably worth running `cargo clippy` locally to catch any other issues, since the build seems to only report one at a time)

Blajda replied:
yeah my bad. I forgot to add the `--tests` flag when running. I'll use the just file going forwards

nicklan reviewed (approved):
awesome, thanks for the contribution! lgtm
